bin/node/cli/tests/websocket_server.rs

```diff
@@ -0,0 +1,427 @@
+#![cfg(not(target_os = "unknown"))]
+#![cfg_attr(docsrs, doc(cfg(all(feature = "std", not(target_os = "unknown")))))]
+
+use async_std::net::{TcpListener, TcpStream};
+use core::{fmt, pin::Pin, str};
+use futures::{channel::mpsc, prelude::*};
+use soketto::handshake::{server::Response, Server};
+use std::{io, net::SocketAddr};
+
+/// Configuration for a [`WsServer`].
+pub struct Config {
+	/// IP address to try to bind to.
+	pub bind_address: SocketAddr,
+
+	/// Maximum size, in bytes, of a frame sent by the remote.
+	///
+	/// Since the messages are entirely buffered before being returned, a maximum value is
+	/// necessary in order to prevent malicious clients from sending huge frames that would
+	/// occupy a lot of memory.
+	pub max_frame_size: usize,
+
+	/// Number of pending messages to buffer up for sending before the socket is considered
+	/// unresponsive.
+	pub send_buffer_len: usize,
+
+	/// Pre-allocated capacity for the list of connections.
+	pub capacity: usize,
+}
+
+/// Identifier for a connection with regard to a [`WsServer`].
+///
+/// After a connection has been closed, its [`ConnectionId`] might be reused.
+#[derive(Debug, Copy, Clone, Ord, PartialOrd, Eq, PartialEq, Hash)]
+pub struct ConnectionId(usize);
+
+/// WebSockets listening socket and list of open connections.
+pub struct WsServer<T> {
+	/// Value passed through [`Config::max_frame_size`].
+	max_frame_size: usize,
+
+	/// Value passed through [`Config::send_buffer_len`].
+	send_buffer_len: usize,
+
+	/// Endpoint for incoming TCP sockets.
+	listener: TcpListener,
+
+	/// Pending incoming connection to accept. Accepted by calling [`WsServer::accept`].
+	pending_incoming: Option<TcpStream>,
+
+	/// List of TCP connections that are currently negotiating the WebSocket handshake.
+	///
+	/// The output can be an error if the handshake fails.
+	negotiating: stream::FuturesUnordered<
+		Pin<
+			Box<
+				dyn Future<Output = (ConnectionId, u64, Result<Server<'static, TcpStream>, ()>)>
+					+ Send,
+			>,
+		>,
+	>,
+
+	/// List of streams of incoming messages for all connections.
+	incoming_messages: stream::SelectAll<
+		Pin<Box<dyn Stream<Item = (ConnectionId, u64, Result<String, ()>)> + Send>>,
+	>,
+
+	/// Tasks dedicated to sending messages on connections. One per healthy connection.
+	sending_tasks:
+		stream::FuturesUnordered<Pin<Box<dyn Future<Output = (ConnectionId, u64)> + Send>>>,
+
+	/// List of connections that are either negotiating or open.
+	connections: slab::Slab<Connection<T>>,
```

Comment:
Do we really need this slab crate here?

Reply:
I don't know :sweat_smile: I took the code from @tomaka

Reply:
I replaced it with a HashMap